src/pyprojroot/criterion.py

```diff
@@ -9,35 +9,50 @@
 import typing
 from os import PathLike as _PathLike
 
+
+_PathType = typing.Union[_PathLike, str]
+_CriterionType = typing.Union[
+    typing.Callable[[_PathType], bool],
+    typing.Callable[[_pathlib.Path], bool],
+    _PathType,
+    _pathlib.Path,
+    typing.Iterable[typing.Callable[[_PathType], bool]],
+    typing.Iterable[typing.Callable[[_pathlib.Path], bool]],
+]
+
 # TODO: It would be nice to have a class that encapsulates these checks,
 # so that we can implement methods like |, !, &, ^ operators
 
 # TODO: Refactor in a way that allows creation of reasons
 
 
-def as_root_criterion(criterion) -> typing.Callable:
+def as_root_criterion(
+    criterion: _CriterionType,
```

Comment:
This type hint is quite complicated. Maybe it would be better replaced with something using functools.singledispatch: https://docs.python.org/3/library/functools.html#functools.singledispatch. That's probably a lot for this PR though.

Reply:
But I wouldn't make an alias for this, since it's only used once. The alias should be for the output type.